### PR TITLE
persist: add initial attempt at monitoring metrics

### PR DESCRIPTION
**Pull request by danhhz**
This certainly contains some we won't end up needing and also doesn't
contain some we will, but gotta start somewhere.

### Motivation

  * This PR adds a feature that has not yet been specified.

### Tips for reviewer

Testing by wiring it up to the open loop benchmark and manually inspecting. We'll want unit test coverage at some point, but I've historically found that class of tests to be a higher than usual maintenance burden, so I figure we should wait a bit.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

**Comment by danhhz** [comment]
@aljoscha @ruchirK this should be ready for a first look! ignore the first commit, that's another PR. I'll fill in the help text once the basic structure has settled down

**Review by aljoscha — COMMENTED** [review]
I like it!

The structure somewhat forces you to add the right metrics fields when adding a new command, but I also thing we'll probably not add to many new commands.

I don't know if where properly instrumenting all the places that we could. For example, not all places that call `retry.sleep()` increment a metric. But I'm not yet sure which of these actually do need to update metrics.

**Comment by danhhz** [comment]
> I don't know if where properly instrumenting all the places that we could. For example, not all places that call retry.sleep() increment a metric. But I'm not yet sure which of these actually do need to update metrics.

yeah, I was trying to be selective, but as long as we're using counters and not histograms, these things are cheap enough that maybe we should just start by erring on the side of more coverage

**Comment by danhhz** [comment]
Okay, had to rebase to resolve a bunch of conflicts and pick stuff up, so sadly this had to be a force-push. Added a `MetricsRetryStream` to help make the RetryStream stuff more failsafe. Plumbed one more of them, but left 3 as TODOs because the plumbing was annoying. I'll take another swing at them tomorrow but they may have to stay as TODOs for now :(

**Comment by danhhz** [comment]
@aljoscha I did 2 of the 3 TODOs I tossed in but I ran out of time for the retry_external one before stopping for the day. IMO this is worthwhile to merge without that and we can revisit after you get back from vacation. Please take a look!

File: src/persist-client/src/impl/metrics.rs [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+
+struct IncOnDrop(IntCounter);
+
+impl Drop for IncOnDrop {
+    fn drop(&mut self) {
+        self.0.inc()
+    }
+}
+
+pub struct MetricsRetryStream {
```
Comment (aljoscha) [inline src/persist-client/src/impl/metrics.rs:264]: Yep, this seems more fool-proof!

**Review by aljoscha — APPROVED** [review]
I think this is good to go! And yes, it's best to merge now, because this is quite big and we don't want it to rot. And we can iterate if need be.

**Comment by aljoscha** [comment]
I poked CI because there were some flakes... :sweat_smile:

**Comment by danhhz** [comment]
TFTR! the retry_external TODO ended up being pretty mechanical, so I took the time to go ahead and just resolve it now. I had a pretty non-trivial rebase (my own fault), so I wanted to double check the manual open-loop testing I'd been doing, anyway

**Pull request merged**